stan/math/prim/mat/prob/neg_binomial_2_log_glm_lpmf.hpp

```diff
@@ -95,107 +124,94 @@ neg_binomial_2_log_glm_lpmf(const T_y& y, const T_x& x, const T_alpha& alpha,
+      = (theta > log_phi)
+            .select(theta + Eigen::log1p(Eigen::exp(log_phi - theta)),
+                    log_phi + Eigen::log1p(Eigen::exp(theta - log_phi)));
+
+  T_sum_val y_plus_phi = y_arr + phi_arr;
 
   // Compute the log-density.
   if (include_summand<propto>::value) {
-    logp -= (y_arr + Array<double, Dynamic, 1>::Ones(N, 1))
-                .unaryExpr(
-                    [](const T_partials_return& xx) { return lgamma(xx); })
-                .sum();
+    logp -= sum(lgamma(y_arr + 1));
   }
   if (include_summand<propto, T_precision>::value) {
-    for (size_t n = 0; n < N; ++n)
-      logp += multiply_log(phi_arr[n], phi_arr[n]) - lgamma(phi_arr[n]);
+    if (is_vector<T_precision>::value) {
+      scalar_seq_view<decltype(phi_val)> phi_vec(phi_val);
+      for (size_t n = 0; n < N; ++n)
+        logp += multiply_log(phi_vec[n], phi_vec[n]) - lgamma(phi_vec[n]);
+    } else {
+      logp += N
+              * (multiply_log(as_scalar(phi_val), as_scalar(phi_val))
+                 - lgamma(as_scalar(phi_val)));
+    }
   }
   if (include_summand<propto, T_x, T_alpha, T_beta, T_precision>::value)
-    logp -= (y_plus_phi * logsumexp_eta_logphi).sum();
+    logp -= (y_plus_phi * logsumexp_theta_logphi).sum();
   if (include_summand<propto, T_x, T_alpha, T_beta>::value)
-    logp += (y_arr * theta_dbl).sum();
+    logp += (y_arr * theta).sum();
   if (include_summand<propto, T_precision>::value) {
-    logp += y_plus_phi
-                .unaryExpr(
-                    [](const T_partials_return& xx) { return lgamma(xx); })
-                .sum();
+    logp += sum(lgamma(y_plus_phi));
   }
 
   // Compute the necessary derivatives.
   operands_and_partials<T_x, T_alpha, T_beta, T_precision> ops_partials(
       x, alpha, beta, phi);
-
   if (!(is_constant_struct<T_x>::value && is_constant_struct<T_beta>::value
-        && is_constant_struct<T_alpha>::value)) {
-    Matrix<T_partials_return, Dynamic, 1> theta_derivative(N, 1);
-    theta_derivative = (y_arr
-                        - (y_plus_phi
-                           / (phi_arr / (theta_dbl.exp())
-                              + Array<double, Dynamic, 1>::Ones(N, 1))))
-                           .matrix();
-    if (!is_constant_struct<T_beta>::value) {
-      ops_partials.edge3_.partials_
-          = value_of(x).transpose() * theta_derivative;
-    }
-    if (!is_constant_struct<T_x>::value) {
-      ops_partials.edge1_.partials_ = theta_derivative * beta_dbl.transpose();
-    }
-    if (!is_constant_struct<T_alpha>::value) {
-      if (is_vector<T_alpha>::value)
-        ops_partials.edge2_.partials_ = theta_derivative;
-      else
-        ops_partials.edge2_.partials_[0] = theta_derivative.sum();
+        && is_constant_struct<T_alpha>::value
+        && is_constant_struct<T_precision>::value)) {
+    Array<T_partials_return, Dynamic, 1> theta_exp = theta.exp();
+    if (!(is_constant_struct<T_x>::value && is_constant_struct<T_beta>::value
```

Comment:
[optional]
It'd be nice if this was just

```
if (!is_constant_struct<T_x, T_beta, T_alpha>::value)
```

because it's hard to read the values we're testing out of all the boilerplate.  But I don't think `is_constant_struct` extends this way.  It'd be easy with parameter packs, but a distraction for this PR.

Reply:
Also added to #1122